mecanum_drive_controller/test/test_mecanum_drive_controller.cpp

```diff
@@ -0,0 +1,547 @@
+{
+  SetUpController();
+
+  ASSERT_EQ(controller_->on_configure(rclcpp_lifecycle::State()), NODE_SUCCESS);
+  ASSERT_EQ(controller_->on_activate(rclcpp_lifecycle::State()), NODE_SUCCESS);
+
+  ASSERT_EQ(
+    controller_->update_reference_from_subscribers(
+      controller_->get_node()->now(), rclcpp::Duration::from_seconds(0.01)),
+    controller_interface::return_type::OK);
+  ASSERT_EQ(
+    controller_->update_and_write_commands(
+      controller_->get_node()->now(), rclcpp::Duration::from_seconds(0.01)),
+    controller_interface::return_type::OK);
+
+  ControllerStateMsg msg;
+  subscribe_and_get_messages(msg);
+
+  EXPECT_FALSE(std::isnan(msg.odom.pose.pose.position.x));
+  // EXPECT_EQ(msg.odom.pose.pose.position.x, 0);
+}
+
+TEST_F(MecanumDriveControllerTest, receive_message_and_publish_updated_status)
+{
+  SetUpController();
+
+  rclcpp::executors::MultiThreadedExecutor executor;
+  executor.add_node(controller_->get_node()->get_node_base_interface());
+
+  ASSERT_EQ(controller_->on_configure(rclcpp_lifecycle::State()), NODE_SUCCESS);
+  ASSERT_EQ(controller_->on_activate(rclcpp_lifecycle::State()), NODE_SUCCESS);
+
+  ASSERT_EQ(
+    controller_->update_reference_from_subscribers(
+      controller_->get_node()->now(), rclcpp::Duration::from_seconds(0.01)),
+    controller_interface::return_type::OK);
+  ASSERT_EQ(
+    controller_->update_and_write_commands(
+      controller_->get_node()->now(), rclcpp::Duration::from_seconds(0.01)),
+    controller_interface::return_type::OK);
+  ControllerStateMsg msg;
+  subscribe_and_get_messages(msg);
+  joint_command_values_[1] = 111;
+  EXPECT_TRUE(std::isnan(msg.linear_x_velocity_command));
+
+  publish_commands(controller_->get_node()->now());
+  ASSERT_TRUE(controller_->wait_for_commands(executor));
+
+  ASSERT_EQ(
+    controller_->update_reference_from_subscribers(
+      controller_->get_node()->now(), rclcpp::Duration::from_seconds(0.01)),
+    controller_interface::return_type::OK);
+  ASSERT_EQ(
+    controller_->update_and_write_commands(
+      controller_->get_node()->now(), rclcpp::Duration::from_seconds(0.01)),
+    controller_interface::return_type::OK);
+  EXPECT_EQ(joint_command_values_[1], 3.0);
+  EXPECT_FALSE(std::isnan(joint_command_values_[1]));
+  fprintf(stderr, " joint_command_values_[1]= %f \n", joint_command_values_[1]);
+
+  subscribe_and_get_messages(msg);
+
+  ASSERT_EQ(msg.linear_x_velocity_command, 1.5);
+}
+
+TEST_F(MecanumDriveControllerTest, test_sending_too_old_message)
+{
+  SetUpController();
+  rclcpp::executors::MultiThreadedExecutor executor;
+  executor.add_node(controller_->get_node()->get_node_base_interface());
+
+  ASSERT_EQ(controller_->on_configure(rclcpp_lifecycle::State()), NODE_SUCCESS);
+  ASSERT_EQ(controller_->on_activate(rclcpp_lifecycle::State()), NODE_SUCCESS);
+
+  // try to set command with time before timeout - command is not updated
+  auto reference = controller_->input_ref_.readFromNonRT();
```

Comment:
You can dereference this already here, then it is easier later.

```suggestion
  auto reference = *(controller_->input_ref_.readFromNonRT());
```